lib/routing.py

```diff
@@ -118,26 +129,79 @@ def redirect(self, path):
         self._dispatch(path)
 
     def _dispatch(self, path):
+        for view_func, rules in iter(self._rules.items()):
+            for rule in rules:
+                if not rule.exact_match(path):
+                    continue
+                log("Dispatching to '%s', exact match" % view_func.__name__)
+                view_func()
+                return
+
+        # then, search for regex matches
         for view_func, rules in iter(self._rules.items()):
             for rule in rules:
                 kwargs = rule.match(path)
-                if kwargs is not None:
-                    log("Dispatching to '%s', args: %s" % (view_func.__name__, kwargs))
-                    view_func(**kwargs)
-                    return
+                if kwargs is None:
+                    continue
+                if self.convert_args:
+                    for k, v in kwargs.items():
+                        new_val = try_convert(v)
+                        if new_val != v:
+                            kwargs[k] = new_val
+                log("Dispatching to '%s', args: %s" % (view_func.__name__, kwargs))
+                view_func(**kwargs)
+                return
         raise RoutingError('No route to path "%s"' % path)
 
 
-class UrlRule(object):
+class Plugin(Addon):
+    """
+    A routing handler bound to a kodi plugin
+    :ivar handle: The plugin handle from kodi
+    :type handle: int
+    """
 
+    def __init__(self, base_url=None, convert_args=False):
+        self.base_url = base_url
+        if self.base_url is None:
+            self.base_url = "plugin://" + xbmcaddon.Addon().getAddonInfo('id')
+        Addon.__init__(self, self.base_url, convert_args)
+        if len(sys.argv) < 2:
+            # we are probably not dealing with a plugin, or it was called incorrectly from an addon
+            raise TypeError('There was no handle provided. This needs to be called from a Kodi Plugin.')
+        self.handle = int(sys.argv[1]) if sys.argv[1].isdigit() else -1
+
+
+class Script(Addon):
+    """
+    A routing handler bound to a kodi script
+    """
```

Comment:
I don't understand why this is needed. "Plugin" is a router (just poorly named), so why can't the router handle any types?

Reply:
Scripts have a different argv. Sometimes ['script.module.fun/do/stuff']. Sometimes ['script.module.fun', '/do/stuff'].
Plugins are always called with ['plugin.video.fun/do/something', '4', '?query=fun'].

If you do routing.Plugin() in a script, it crashes because the argv is different. You could add a handler in __init__(), but I prefer OOP where Plugin is a plugin, therefore assumed to have a handle, while Script does not.